### DISCUSSION TITLE
Using UploadFile and Pydantic model in one request

### DISCUSSION THREAD

**Discussion by vobinics** [fastapi/fastapi#7188] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
from fastapi import FastAPI, File, UploadFile
from pydantic import BaseModel

app = FastAPI()


class Properties(BaseModel):
    language: str = None
    author: str = None


@app.post("/uploadfile/", status_code=201)
async def create_upload_file(properties: Properties, file: UploadFile = File(...)):
    return {"filename": file.filename, 'properties': properties}
```

### Description

* Open the browser `/docs` and call the endpoint `/uploadfile`.
* It returns error 422 Validation Error.
* But I expected it to return code 201.

### Environment

* OS: Linux (Fedora)
* FastAPI Version: 0.61.1
* Python version: 3.8.6

### Additional context
Immediately I apologize, maybe I missed something and / or did not understand and should not have bothered you. I need this function in order to unload a poster for it along with information in the form of a Pydantic model. I issued a [patch](https://github.com/tiangolo/fastapi/pull/2139), but everyone ignores it

**Comment by includeamin** [comment]
based on [docs](https://fastapi.tiangolo.com/tutorial/request-forms-and-files/#import-file-and-form)
``` python
from fastapi import FastAPI, File, UploadFile,Form
from pydantic import BaseModel

app = FastAPI()


class Properties(BaseModel):
    language: str = None
    author: str = None


@app.post("/uploadfile/", status_code=201)
async def create_upload_file(language: str = Form(...),author: str = Form(...),file: UploadFile = File(...)):
    return {"filename": file.filename, 'properties': Properties(language=language,author=author)}
```

**Comment by vobinics** [comment]
> based on [docs](https://fastapi.tiangolo.com/tutorial/request-forms-and-files/#import-file-and-form)
> ``` python
> from fastapi import FastAPI, File, UploadFile,Form
> from pydantic import BaseModel
> 
> app = FastAPI()
> 
> 
> class Properties(BaseModel):
>     language: str = None
>     author: str = None
> 
> 
> @app.post("/uploadfile/", status_code=201)
> async def create_upload_file(language: str = Form(...),author: str = Form(...),file: UploadFile = File(...)):
>     return {"filename": file.filename, 'properties': Properties(language=language,author=author)}
> ```

Hello. This option does not suit me, since the model in my project has many fields and I do not want crutches in my project, because of this I released a patch so that you can transfer the Pydantic Model to FormData. But thanks for trying to help

**Comment by SirTelemak** [comment]
Then you can try to use dependency injection. Just make class with required fields (better use dataclass). 
But imo it's bad practice to upload file and body fields simultaneously

**Comment by vobinics** [comment]
> Then you can try to use dependency injection. Just make class with required fields (better use dataclass).

Again, this will be a crutch, I already have a ready-made Pydantic model and a handler for it. Now I just want to add a picture to it. FastAPI does not support the Pydantic model through FormData. Why can't you just add support for this? I already solved this problem with 4 lines of code but they are ignored. But thanks for trying to help

**Comment by SirTelemak** [comment]
> Again, this will be a crutch, I already have a ready-made Pydantic model and a handler for it. Now I just want to add a picture to it. FastAPI does not support the Pydantic model through FormData. Why can't you just add support for this? I already solved this problem with 4 lines of code but they are ignored. But thanks for trying to help

In fact, it's working with Pydantic model, I had such code snippet in my pet project:
```python3
@api.post('/')
def add_review(
    body: Model = Body(...),
    image_files: Optional[List[UploadFile]] = File(None, media_type='image/jpeg'),
):
    ...
```
BUT! when I tried to send such data I had to insert body json as string inside form-data (that's RFC limitations: https://tools.ietf.org/html/rfc1867)

**Comment by vobinics** [comment]
> In fact, it's working with Pydantic model, I had such code snippet in my pet project:
> 
> ```python
> @api.post('/')
> def add_review(
>     body: Model = Body(...),
>     image_files: Optional[List[UploadFile]] = File(None, media_type='image/jpeg'),
> ):
>     ...
> ```
> 
> BUT! when I tried to send such data I had to insert body json as string inside form-data (that's RFC limitations: https://tools.ietf.org/html/rfc1867)

The code you provided doesn't work ("Model" in your example inherits the Pydantic model right?) (At least with Swagger). Error 422 (Swagger still submits the Pydantic model via form data and this cannot be handled by FastAPI). And if this does not work in Swagger, then I cannot make tests for my project (pytest)

**Comment by vobinics** [comment]
```Python
@app.post("/uploadfile", status_code=201)
async def create_upload_file(properties: Properties = Form(...), file: UploadFile = File(...)):
    return {"filename": file.filename, 'properties': properties}
```

We left on the wrong topic. I want to discuss why I cannot use FormData

**Comment by SirTelemak** [comment]
Oups, sorry, I forgot I made custom validator to transform str to json for Model:
```python3
class Model(BaseModel):
    foo: int
    bar: str

    @classmethod
    def __get_validators__(cls):
        yield cls.validate_to_json

    @classmethod
    def validate_to_json(cls, value):
        if isinstance(value, str):
            return cls(**json.loads(value))
        return value
```

**Comment by vobinics** [comment]
> Oups, sorry, I forgot I made custom validator to transform str to json for Model:
> 
> ```python
> class Model(BaseModel):
>     foo: int
>     bar: str
> 
>     @classmethod
>     def __get_validators__(cls):
>         yield cls.validate_to_json
> 
>     @classmethod
>     def validate_to_json(cls, value):
>         if isinstance(value, str):
>             return cls(**json.loads(value))
>         return value
> ```

I've seen a similar solution, but you don't think it is a workaround. Maybe it's time to really fix the problem? You can try using my fork with the problem fixed

**Comment by SirTelemak** [comment]
> We left on the wrong topic. I want to discuss why I cannot use FormData

I can't give you answer, but I think it's because Form data is string and cannot be parsed easily. For example, you cannot use Pydantic model in Query too, only for Body params

**Comment by vobinics** [comment]
> > Мы ушли не по той теме. Я хочу обсудить, почему я не могу использовать FormData
> 
> Я не могу дать вам ответ, но думаю, это потому, что данные формы являются строковыми и не могут быть легко проанализированы. Например, вы не можете использовать модель Pydantic в Query, только для параметров Body.

[Open it ](https://github.com/tiangolo/fastapi/pull/2139)

**Comment by SirTelemak** [comment]
Imo it's still workaround and even dirtier than custom validator. Form Data used for text and files commonly, not for json data and making it as default behavior could be misleading and make more bugs. 
As I said, look at Form like at Query, they share identical behavior - some small atomic data
Still, it could be my stubbornness and I'd like to hear another one opinion

**Comment by vobinics** [comment]
> Imo it's still workaround and even dirtier than custom validator. Form Data used for text and files commonly, not for json data and making it as default behavior could be misleading and make more bugs.
> As I said, look at Form like at Query, they share identical behavior - some small atomic data
> Still, it could be my stubbornness and I'd like to hear another one opinion

Perhaps you are right, we will wait for opinions on this from other participants.

**Comment by vobinics** [comment]
A small digression. I am currently switching from Django there it uses Form to pass new record data. Why the same cannot be done in FastAPI. It's just a question

**Comment by Freestyle1190** [comment]
Hi,

if I understand you well you can use UploadFile and Pydantic model in one request using **Dependencies**.
More detailed information here: [https://fastapi.tiangolo.com/tutorial/dependencies/](url)
Sample code:
```python
from fastapi import File, UploadFile, Depends

class User(BaseModel):

    username: str
    password: str

@app.post('/user')
async def upload(user: User = Depends(), file: UploadFile = File(...)):

  data_db = user.dict()
  print(data_db)
```

**Comment by vobinics** [comment]
> Hi,
> 
> if I understand you well you can use UploadFile and Pydantic model in one request using **Dependencies**.
> More detailed information here: [https://fastapi.tiangolo.com/tutorial/dependencies/](url)
> Sample code:
> 
> ```python
> from fastapi import File, UploadFile, Depends
> 
> class User(BaseModel):
> 
>     username: str
>     password: str
> 
> @app.post('/user')
> async def upload(user: User = Depends(), file: UploadFile = File(...)):
> 
>   data_db = user.dict()
>   print(data_db)
> ```

Hello. Thanks for trying to help, but that's not what I want. My pidantic model is quite large and it's not very good to use GET parameters for it. I think it is better to use custom validator in pydantic model and pass data through FormData. I think it's time to close this topic

**Comment by vobinics** [comment]
> Oups, sorry, I forgot I made custom validator to transform str to json for Model:
> 
> ```python
> class Model(BaseModel):
>     foo: int
>     bar: str
> 
>     @classmethod
>     def __get_validators__(cls):
>         yield cls.validate_to_json
> 
>     @classmethod
>     def validate_to_json(cls, value):
>         if isinstance(value, str):
>             return cls(**json.loads(value))
>         return value
> ```

This is the solution to the problem. But it would be better if you can use the pydantic model through FormData without additional code. There is a ready-made [patch](https://github.com/tiangolo/fastapi/pull/2139) for this problem

**Reply by aelamran** [reply]
__get_validators__ is this working with pydantic V2, I am afraid it is not.

Also curious to know your concerns about:

> Hello. Thanks for trying to help, but that's not what I want. My pidantic model is quite large and it's not very good to use GET parameters for it. I think it is better to use custom validator in pydantic model and pass data through FormData. I think it's time to close this topic

**Comment by tiangolo** [comment]
Thanks for the help here everyone! :clap: :bow: 

Thanks for reporting back and closing the issue :+1:

**Comment by beishangongzi** [comment]
I am a new web developer. Is it better to have two requests for pydantic and uploadFile or one request?

**Comment by Farshadkouhkan** [comment]
hello guys i would send data from form data and upload image in the same time and validate the image inside the Base Model how can i do ?

**Comment by a1k0u** [comment]
> hello guys i would send data from form data and upload image in the same time and validate the image inside the Base Model how can i do ?

You're able to send your file in binary form (base64 will help to convert from string to binary, check out docs), if you want use BaseModel for files. From my point of view it isn't good idea :)

**Comment by Farshadkouhkan** [comment]
You can use from file chunks for validate

**Comment by tiangolo** [comment]
Hello there! If you have new questions, it's better to ask them in new issues, following all the template, creating a code example that shows what you want to achieve, etc.

**Comment by nathanneibauer** [comment]
I am having the same issue today from @vobinics original post. I am trying to pass a file _and_ metadata, all as form data. Like @vobinics , I have many fields and want to use a pydantic model instead of passing multiple `Form(...)` parameters. Has anyone successfully figured this out?

**Reply by johannesmik** [reply]
Yes, a working solution, already discussed in this thread, is to add a custom validator to your pydantic model:

> > Oups, sorry, I forgot I made custom validator to transform str to json for Model:
> > ```python
> > class Model(BaseModel):
> >     foo: int
> >     bar: str
> > 
> >     @classmethod
> >     def __get_validators__(cls):
> >         yield cls.validate_to_json
> > 
> >     @classmethod
> >     def validate_to_json(cls, value):
> >         if isinstance(value, str):
> >             return cls(**json.loads(value))
> >         return value
> > ```
> 
> This is the solution to the problem. But it would be better if you can use the pydantic model through FormData without additional code. There is a ready-made [patch](https://github.com/tiangolo/fastapi/pull/2139) for this problem

**Reply by mahadi** [reply]
According to this comment https://github.com/fastapi/fastapi/pull/2139#issuecomment-2336825724 it should be implemented in 0.113 but i cannot get it working. As soon as i add the files i get `{'detail': [{'type': 'missing', 'loc': ['body', 'data'], 'msg': 'Field required', 'input': None}]}`, where data is the name of my pydantic model variable.

Also the custom validator doesn't make any difference:

``` python
class FormData(BaseModel):
    username: str
    password: str

    @classmethod
    def __get_validators__(cls):
        yield cls.validate_to_json

    @classmethod
    def validate_to_json(cls, value):
        if isinstance(value, str):
            return cls(**json.loads(value))
        return value

# ...
@app.post('/login')
def login(data: FormData = Form(...), files: List[UploadFile] = File(...)):
    return data, files


# ...
payload = {
    'username': 'admin',
    'password': 'pwd',
}

files = [
    ('files', open(r"D:/temp/a.txt", 'rb')),
]

resp = requests.post(
    url=f"http://localhost:{PORT}/login/",
    data=payload,
    files=files,
)

print(resp.json())
```

Is there any obvious mistake?

**Comment by YuriiMotov** [comment, marked as answer]
https://github.com/fastapi/fastapi/pull/2139#issuecomment-2336825724

> Thanks! ☕
> 
> This was covered in #12129, it is available in FastAPI 0.113.0 🎉
> 
> The new docs: https://fastapi.tiangolo.com/tutorial/request-form-models/
> 
> Given that, I'll close this one, but thanks for the effort! 🍰

Since 0.113.0 FastAPI supports Pydantic models to declare Form parameters 🎉
Docs: https://fastapi.tiangolo.com/tutorial/request-form-models/
